src/sage/modular/hypergeometric_motive.py

```diff
@@ -146,33 +153,42 @@ def characteristic_polynomial_from_traces(traces, d, q, i, sign):
         sage: characteristic_polynomial_from_traces([22, 484], 4, 31, 2, -1)
         -923521*T^4 + 21142*T^3 - 22*T + 1
 
+        sage: characteristic_polynomial_from_traces([22], 4, 31, 2, -1, deg=1)
+        -22*T + 1
+        sage: characteristic_polynomial_from_traces([22, 484], 4, 31, 2, -1, deg=4)
+        -923521*T^4 + 21142*T^3 - 22*T + 1
+
     TESTS::
 
         sage: characteristic_polynomial_from_traces([-36], 4, 17, 3, 1)
         Traceback (most recent call last):
         ...
         ValueError: not enough traces were given
     """
-    if len(traces) < d // 2:
+    if use_fe:
+        bound = d // 2 if deg is None else min(d // 2, deg)
+    else:
+        bound = deg
+    if len(traces) < bound:
         raise ValueError('not enough traces were given')
-    if i % 2 and d % 2:
+    if i % 2 and d % 2 and use_fe:
         raise ValueError('i and d may not both be odd')
     t = PowerSeriesRing(QQ, 't').gen()
     ring = PolynomialRing(ZZ, 'T')
 
     series = sum(- api * t**(i + 1) / (i + 1) for i, api in enumerate(traces))
-    series = series.O(d // 2 + 1).exp()
+    series = series.O(bound + 1).exp()
     coeffs = list(series)
-    coeffs += [0] * max(0, d // 2 + 1 - len(coeffs))
+    coeffs += [0] * max(0, bound + 1 - len(coeffs))
 
-    data = [0 for _ in range(d + 1)]
-    for k in range(d // 2 + 1):
+    fulldeg = d if deg is None else deg
+    data = [0 for _ in range(fulldeg + 1)]
```

Comment:
```suggestion
    data = [0] * (fulldeg + 1)
```
Might be slightly faster, but at least IMO it is easier to read.